percona/tests/upstream_update/performance_test.go

```diff
@@ -0,0 +1,217 @@
+		if !assert.NoError(t, err) {
+			return nil
+		}
+
+		durations = append(durations, float64(d))
+		hwms = append(hwms, float64(hwm))
+		datas = append(datas, float64(data))
+	}
+
+	mean, _ := stats.Mean(durations)
+	stdDev, _ := stats.StandardDeviation(durations)
+	stdDev = float64(100) / mean * stdDev
+
+	clockTicks, err := sysconf.Sysconf(sysconf.SC_CLK_TCK)
+	if err != nil {
+		panic(err)
+	}
+
+	mean = mean * float64(1000) / float64(clockTicks) / float64(size)
+	stdDevMs := stdDev / float64(100) * mean
+
+	meanHwm, _ := stats.Mean(hwms)
+	stdDevHwm, _ := stats.StandardDeviation(hwms)
+	stdDevHwmPerc := float64(100) / meanHwm * stdDevHwm
+
+	meanData, _ := stats.Mean(datas)
+	stdDevData, _ := stats.StandardDeviation(datas)
+	stdDevDataPerc := float64(100) / meanData * stdDevData
+
+	st := StatsData{
+		meanMs:     mean,
+		stdDevMs:   stdDevMs,
+		stdDevPerc: stdDev,
+
+		meanHwm:        meanHwm,
+		stdDevHwmBytes: stdDevHwm,
+		stdDevHwmPerc:  stdDevHwmPerc,
+
+		meanData:        meanData,
+		stdDevDataBytes: stdDevData,
+		stdDevDataPerc:  stdDevDataPerc,
+	}
+
+	//fmt.Printf("loop %dx%d: sample time: %.2fms [deviation ±%.2fms, %.1f%%]\n", cnt, scrapesCount, st.meanMs, st.stdDevMs, st.stdDevPerc)
+	fmt.Printf("running %d scrapes %d times\n", size, cnt)
+	fmt.Printf("CPU\t%.1fms [±%.1fms, %.1f%%]\n", st.meanMs, st.stdDevMs, st.stdDevPerc)
+	fmt.Printf("HWM\t%.1fkB [±%.1f kB, %.1f%%]\n", st.meanHwm, st.stdDevHwmBytes, st.stdDevHwmPerc)
+	fmt.Printf("Data\t%.1fkB [±%.1f kB, %.1f%%]\n", st.meanData, st.stdDevDataBytes, st.stdDevDataPerc)
+
+	return &st
+}
+
+func doTest(iterations int, fileName string) (cpu, hwm, data int64, _ error) {
+	cmd, port, collectOutput, err := launchExporter(fileName)
+	if err != nil {
+		return 0, 0, 0, err
+	}
+
+	total1 := getCPUTime(cmd.Process.Pid)
+
+	for i := 0; i < iterations; i++ {
+		_, err = tryGetMetrics(port)
+		if err != nil {
+			return 0, 0, 0, errors.Wrapf(err, "Failed to perform test iteration %d.%s", i, collectOutput())
+		}
+
+		time.Sleep(1 * time.Millisecond)
+	}
+
+	total2 := getCPUTime(cmd.Process.Pid)
+
+	hwm, data = getCPUMem(cmd.Process.Pid)
+
+	err = stopExporter(cmd, collectOutput)
+	if err != nil {
+		return 0, 0, 0, err
+	}
+
+	return total2 - total1, hwm, data, nil
+}
+
+func getCPUMem(pid int) (hwm, data int64) {
+	contents, err := ioutil.ReadFile(fmt.Sprintf("/proc/%d/status", pid))
+	if err != nil {
+		return 0, 0
+	}
+
+	lines := strings.Split(string(contents), "\n")
+
+	for _, v := range lines {
+		if strings.HasPrefix(v, "VmHWM") {
+			val := strings.ReplaceAll(strings.ReplaceAll(strings.Split(v, ":\t")[1], " kB", ""), " ", "")
+			hwm, _ = strconv.ParseInt(val, 10, 64)
+			continue
+		}
+		if strings.HasPrefix(v, "VmData") {
+			val := strings.ReplaceAll(strings.ReplaceAll(strings.Split(v, ":\t")[1], " kB", ""), " ", "")
+			data, _ = strconv.ParseInt(val, 10, 64)
+			continue
+		}
+	}
+
+	return hwm, data
+}
+
+func getCPUTime(pid int) (total int64) {
```

Comment:
Again, it seems that there is no significant reason to use named return value here. Better to avoid them: https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_named_return_values